pkg/block/gs/adapter.go

```diff
@@ -85,14 +101,68 @@ func (a *Adapter) newPreSignedTime() time.Time {
 	return time.Now().UTC().Add(a.preSignedExpiry)
 }
 
+/*
+Ideally we assume all the objects should be encrypted with AES key at the very beginning
+It may be that some existing objects were created before the key is was introduced, so objects will be examined and checked for encryption.
+Multiple keys for individual objects aren't supported! An error will be generated if an improper key is supplied.
+Keys in KMS will not be validated. You should use a proper service account with permissions to access them.
+*/
+func (o *storageObjectHandle) prepareReadHandle(ctx context.Context, a *Adapter) *storage.ObjectHandle {
+	att, err := o.Attrs(ctx)
+	if err == nil {
+		a.log(ctx).Debug("object has attribute customerKeySHA256 means it is encrypted by Customer-Supplied key: ", att.CustomerKeySHA256)
+		if a.ServerSideEncryptionCustomerSupplied != nil && att.CustomerKeySHA256 != "" {
+			return o.Key(a.ServerSideEncryptionCustomerSupplied)
+		}
+	}
+	// Assume no decryption needed when attrs is not found
+	a.log(ctx).Debugf("object Attrs get error %w", err)
+	return o.ObjectHandle
+}
+
+type storageObjectHandle struct {
+	*storage.ObjectHandle
+}
+
+func (o *storageObjectHandle) prepareWriteHandle(a *Adapter) *storageObjectHandle {
```

Comment:
This is classic for the `with` prefix:
```suggestion
func (o *storageObjectHandle) withWriteHandle(a *Adapter) *storageObjectHandle {
```